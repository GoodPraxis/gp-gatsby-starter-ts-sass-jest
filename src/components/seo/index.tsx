/**
 * SEO component that queries for data with
 *  Gatsby's useStaticQuery React hook
 *
 * See: https://www.gatsbyjs.org/docs/use-static-query/
 */

import React from 'react';
import { Helmet } from 'react-helmet';
import { useStaticQuery, graphql } from 'gatsby';

interface SEOProps {
  description?: string;
  lang?: string;
  meta?: { name: string; content: any; property?: undefined; }[];
  image?: string;
  title?: string;
  metaTitle?: string;
  metaAuthor?: string;
}

export function getMeta({
  description = '',
  meta = [],
  title = '',
  metaTitle = '',
  image = '',
}: SEOProps) {
  return [
    {
      name: 'description',
      content: description,
    },
    {
      property: 'og:image',
      content: image,
    },
    {
      property: 'og:title',
      content: metaTitle,
    },
    {
      property: 'og:description',
      content: description,
    },
    {
      property: 'og:type',
      content: 'website',
    },
    {
      name: 'twitter:card',
      content: 'summary',
    },
    {
      name: 'twitter:title',
      content: title,
    },
    {
      name: 'twitter:description',
      content: description,
    },
  ].concat(meta);
}

export const PureSEO: React.FC<SEOProps> = (props: SEOProps) => {
  const { lang = 'en', title } = props;
  return (
    <Helmet
      htmlAttributes={{
        lang,
      }}
      title={title}
      titleTemplate="%s"
      meta={getMeta(props)}
    />
  );
};

/*
  istanbul ignore next
  Covered in E2E test
*/
const SEO: React.FC<SEOProps> = ({
  description = '', lang = 'en', meta = [], title,
}: SEOProps) => {
  const { site } = useStaticQuery(
    graphql`
      query {
        site {
          siteMetadata {
            title
            description
            author
            image
          }
        }
      }
    `,
  );

  const metaDescription = description || site.siteMetadata.description;
  const { title: metaTitle, metaAuthor, image } = site.siteMetadata;

  return (
    <PureSEO
      title={title}
      metaTitle={metaTitle}
      description={metaDescription}
      lang={lang}
      metaAuthor={metaAuthor}
      meta={meta}
      image={image}
    />
  );
};

export default SEO;
