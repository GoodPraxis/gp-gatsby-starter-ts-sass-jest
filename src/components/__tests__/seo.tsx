import renderer from 'react-test-renderer';
import React from 'react';
import { getMeta, PureSEO as SEO } from '../seo';

describe('SEO', () => {
  it('creates correct SEO props', () => {
    const meta = getMeta({
      description: 'Foobar description',
      title: 'FooBar',
      metaTitle: 'FooBar',
      image: 'foo.bar',
    });
    expect(meta).toMatchSnapshot();
  });

  it('renders the component without any errors', () => {
    renderer.create(<SEO />);
    renderer.create(<SEO title="foobar" />);
    renderer.create(<SEO meta={[{ name: 'foo', content: 'bar' }]} />);
  });

  it('concatenates meta data', () => {
    const meta = getMeta({
      meta: [{ name: 'foo', content: 'bar' }],
    });
    expect(meta.filter(
      (e) => e.name === 'foo' && e.content === 'bar',
    ).length).toBe(1);
  });
});
