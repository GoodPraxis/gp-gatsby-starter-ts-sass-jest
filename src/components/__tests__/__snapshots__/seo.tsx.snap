// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SEO creates correct SEO props 1`] = `
Array [
  Object {
    "content": "Foobar description",
    "name": "description",
  },
  Object {
    "content": "foo.bar",
    "property": "og:image",
  },
  Object {
    "content": "FooBar",
    "property": "og:title",
  },
  Object {
    "content": "Foobar description",
    "property": "og:description",
  },
  Object {
    "content": "website",
    "property": "og:type",
  },
  Object {
    "content": "summary",
    "name": "twitter:card",
  },
  Object {
    "content": "FooBar",
    "name": "twitter:title",
  },
  Object {
    "content": "Foobar description",
    "name": "twitter:description",
  },
]
`;
